File: kafka-demo-java11/src/main/java/com/example/streamfunctiondemo/controller/KafkaDemoController.java
package com.example.streamfunctiondemo.controller;

import com.example.streamfunctiondemo.repository.BatchInsertRepository;
import com.example.streamfunctiondemo.repository.Person;
import com.example.streamfunctiondemo.repository.PersonRepository;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.commons.lang.RandomStringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cloud.stream.function.StreamBridge;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.concurrent.ForkJoinPool;

@RestController
public class KafkaDemoController {

    @Autowired
    private StreamBridge streamBridge;

    @Autowired
    private PersonRepository personRepository;

    @Autowired
    private BatchInsertRepository batchInsertRepository;

    private final String myDestination = "mytopic";

    private static String[] genderFlag = {"M","F"};
    /**
     * 10000개의 랜덤한 String을 생성하여 Topic에 publish한다
     */
    @GetMapping("/test/{maxCount}")
    public void produceTest(@PathVariable int maxCount){
        long start = System.currentTimeMillis();
        this.personGenerator(maxCount)
                .stream().parallel()
                .forEach(it->{
//                    System.out.println("Thread : "+Thread.currentThread().getName()+", value : " + it.getName());
                    this.streamBridge.send(myDestination,it);
                });
        System.out.println("api exec time : "+(System.currentTimeMillis()-start));
    }

    private List<Person> personGenerator(int maxCnt){
        long start =System.currentTimeMillis();
        List<Person> res = new ArrayList<>();

        Random rand = new Random();
        for(int i=0; i<maxCnt; i++){
            //name
            String randName = RandomStringUtils.randomAlphanumeric(7).toUpperCase();

            //gender
            String randGender = genderFlag[(rand.nextInt(2))];

            //phoneNumber
            String phoneNumber = "010".concat(RandomStringUtils.random(8,33,125,false,true));


            res.add(Person.of(randName,randGender,rand.nextInt(100),phoneNumber));
        }
        System.out.println("generator exec time : " + (System.currentTimeMillis()-start));
        return res;
    }


    @PostMapping("/resttest")
    public void receiveMessageApi(@RequestBody Person person){
        ObjectMapper objectMapper = new ObjectMapper();

        try{
            System.out.println("===> Recieved Message : "+ objectMapper.writeValueAsString(person));
            this.personRepository.save(person);
        }catch(Exception e){
            System.out.println("======Error=====");
            System.out.println(e.getMessage());
        }
    }

}

File: kafka-demo-kotlin/src/main/kotlin/com/example/kafkademo/kafkademo/KafkaDemoController.kt
/**
 * ===============================================================
 * File name : KafkaDeomController.java
 * Created by injeahwang on 2021-04-02
 * ===============================================================
 */
package com.example.kafkademo.kafkademo

import kotlinx.coroutines.delay
import kotlinx.coroutines.runBlocking
import org.apache.commons.lang.RandomStringUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.cloud.function.json.GsonMapper
import org.springframework.cloud.function.json.JacksonMapper
import org.springframework.cloud.stream.function.StreamBridge
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.HttpStatus
import org.springframework.integration.support.json.JacksonJsonUtils
import org.springframework.messaging.support.GenericMessage
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController
import reactor.core.publisher.Flux
import reactor.core.publisher.Sinks
import reactor.util.concurrent.Queues
import java.util.*
import java.util.function.Consumer
import java.util.function.Supplier

@RestController
@Configuration
class KafkaDemoController {
    @Autowired
    private val streamBridge: StreamBridge? = null

    //    @Qualifier("publishTest")
    private val emitter = Sinks.many().multicast().onBackpressureBuffer<SendObj>(Queues.SMALL_BUFFER_SIZE, true)

    //    @Autowired
    //    @Qualifier("publishTest")
    //    private Supplier<Flux<String>> publishTest;
    private val myDestination = "mytopic"

    /**
     * 10000개의 랜덤한 String을 생성하여 Topic에 publish한다
     * 1000개의요청이 100ms 대기
     */
    @GetMapping("/v1/test")
    @ResponseStatus(HttpStatus.ACCEPTED)
    fun produceTest() {

        val start = System.currentTimeMillis()
        for( i in 0 until 10){
            runBlocking {
                stringListGenerator(1000)
                        .forEach(Consumer { t: SendObj -> emitter.tryEmitNext(t) })
                delay(100L)
            }
        }

        println("api lead time : " + (System.currentTimeMillis() - start))
    }

    private fun stringListGenerator(maxCnt: Int): List<SendObj> {
        val start = System.currentTimeMillis()
        val res: MutableList<SendObj> = ArrayList()
        for (i in 0 until maxCnt) {
            val newStr = RandomStringUtils.randomAlphanumeric(17).toUpperCase()
            res.add(SendObj(value=newStr))
        }
        println("generator lead time : " + (System.currentTimeMillis() - start))
        return res
    }

    @Bean
    fun consumeTest(): Supplier<Flux<SendObj>> {
        return Supplier { emitter.asFlux() }
    }
}

data class SendObj(
//        val id: Long?,
        val value: String
)